CORS-3437: infra/capi: add provisioning timeout

patrickdillon opened the pull request:
Implements basic safeguard so that provisioning does not spin indefinitely. A timeout (currently 15m) is set for each provisioning stage.

patrickdillon commented:
/test altinfra-e2e-aws-ovn altinfra-e2e-azure-capi-ovn altinfra-e2e-nutanix-capi-ovn altinfra-e2e-vsphere-capi-ovn altinfra-e2e-openstack-capi-ovn altinfra-e2e-gcp-capi-ovn

patrickdillon commented on pkg/infrastructure/clusterapi/clusterapi.go:
```diff
@@ -108,7 +111,9 @@ func (i *InfraProvider) Provision(ctx context.Context, dir string, parents asset
 		}
 
-		if err := p.PreProvision(ctx, preProvisionInput); err != nil {
+		preCtx, cancel := context.WithTimeout(ctx, timeout)
+		defer cancel()
+		if err := p.PreProvision(preCtx, preProvisionInput); err != nil {
```
I'm going to remove these timeouts from the hooks and only keep the timeout caps on the provisioning with the capi system. We don't need to handle hook failures with timeouts, instead we can leave that up to the hook implementors. For CAPI provisioning on the other hand, we need a timeout to prevent endlessly spinning.

patrickdillon commented:
/test altinfra-e2e-aws-ovn altinfra-e2e-azure-capi-ovn altinfra-e2e-nutanix-capi-ovn altinfra-e2e-vsphere-capi-ovn altinfra-e2e-openstack-capi-ovn altinfra-e2e-gcp-capi-ovn

r4f4 commented:
/cc

r4f4 commented on pkg/infrastructure/clusterapi/clusterapi.go:
```diff
@@ -173,7 +190,7 @@ func (i *InfraProvider) Provision(ctx context.Context, dir string, parents asset
 			return cluster.Status.InfrastructureReady, nil
 		}); err != nil {
-			return fileList, err
+			return fileList, fmt.Errorf("infrastructure was not ready within %v: %w", timeout, err)
```
What about:
```suggestion
            if wait.Interrupted(err) {
			    return fileList, fmt.Errorf("infrastructure was not ready within %v", timeout)
			}
			return fileList, err
```

r4f4 replied:
Otherwise it reads like we're always timing out even when we fail immediately?

patrickdillon replied:
perfect, thanks!

r4f4 commented on pkg/infrastructure/clusterapi/clusterapi.go:
```diff
@@ -266,9 +290,10 @@ func (i *InfraProvider) Provision(ctx context.Context, dir string, parents asset
 			return true, nil
 		}); err != nil {
-			return fileList, err
+			return fileList, fmt.Errorf("machines were not provisioned within %v: %w", timeout, err)
```
Same as above.

r4f4 reviewed (commented):
How do you feel about adding
```
               untilTime := time.Now().Add(timeout)
               timezone, _ := untilTime.Zone()
               logrus.Infof("Waiting up to %v (until %v %s) for infrastructure to provision...", timeout, untilTime.Format(time.Kitchen), timezone)
```
Too much information?

patrickdillon commented:
/test altinfra-e2e-aws-ovn

patrickdillon commented:
All comments incorporated.

r4f4 commented on pkg/infrastructure/clusterapi/clusterapi.go:
```diff
@@ -40,6 +41,19 @@ import (
+	// timeout for each provisioning step.
+	timeout = 15 * time.Minute
+
+	// infrastructure provisioning stages
```
```suggestion
	// infrastructure provisioning stages.
```
to make the linter happy.

patrickdillon replied:
doh, I forgot to delete that with my last changes. pushed.

r4f4 reviewed (commented):
LGTM. Will wait for the linting fix before tagging.

r4f4 commented:
Looking good: 
```log
time="2024-04-25T20:44:52Z" level=info msg="Waiting up to 15m0s (until 8:59PM UTC) for infrastructure to become ready..."
[...]
time="2024-04-25T20:50:14Z" level=info msg="Creating private Hosted Zone"
[...]
time="2024-04-25T20:50:45Z" level=info msg="Waiting up to 15m0s (until 9:05PM UTC) for machines to provision..."
[...]
time="2024-04-25T20:51:10Z" level=debug msg="Machine ci-op-fnm19l75-2d061-p5kqn-master-0 is ready. Phase: Provisioned"
time="2024-04-25T20:51:10Z" level=debug msg="Machine ci-op-fnm19l75-2d061-p5kqn-master-1 is ready. Phase: Provisioned"
time="2024-04-25T20:51:10Z" level=debug msg="Machine ci-op-fnm19l75-2d061-p5kqn-master-2 is ready. Phase: Provisioned"
time="2024-04-25T20:51:10Z" level=info msg="Cluster API resources have been created. Waiting for cluster to become ready..."
[...]
time="2024-04-25T21:18:38Z" level=debug msg="Time elapsed per stage:"
time="2024-04-25T21:18:38Z" level=debug msg="     Infrastructure Provisioning: 5m29s"
time="2024-04-25T21:18:38Z" level=debug msg="InfrastructureReady Provisioning: 31s"
time="2024-04-25T21:18:38Z" level=debug msg="            Machine Provisioning: 26s"
time="2024-04-25T21:18:38Z" level=debug msg="              Bootstrap Complete: 12m20s"
time="2024-04-25T21:18:38Z" level=debug msg="                             API: 5m41s"
time="2024-04-25T21:18:38Z" level=debug msg="               Bootstrap Destroy: 2m1s"
time="2024-04-25T21:18:38Z" level=debug msg="     Cluster Operators Available: 9m8s"
time="2024-04-25T21:18:38Z" level=debug msg="        Cluster Operators Stable: 3m58s"
time="2024-04-25T21:18:38Z" level=info msg="Time elapsed: 34m4s"
```

patrickdillon commented:
fixed linter and reworked stage names a little: "InfrastructureReady" would not be obvious to users. 

/test altinfra-e2e-aws-ovn altinfra-e2e-azure-capi-ovn altinfra-e2e-nutanix-capi-ovn altinfra-e2e-vsphere-capi-ovn altinfra-e2e-openstack-capi-ovn altinfra-e2e-gcp-capi-ovn

r4f4 reviewed (commented):
/approve
/lgtm
